### ISSUE TITLE
Major drop in performance and speed

### ISSUE THREAD

**Issue by jbmelander** [flatironinstitute/mountainsort5#46]
My lab members were experiencing some serious challenges spikesorting using code and datasets that I had been quite successful with. Those challenges mainly looked like ms5 taking ~4x to run, and getting garbage units that failed our most basic quality-controls. After much debugging, I discovered that it was because the newest spikeinterface release somehow makes the code extremely slow and ineffective. I can run the same code with spikeinterface (version 0.99.1) and get great units with quality receptive-fields in 20 minutes. Then, upgrading to the newest spikeinterface (as the only change in the pipelien) it will take > 3 hours and will get a single unit containing noise.

A few things:
1. Do you have any insight about why this might be happening?
2. This also means that the neuropixel quickstart example will no longer work if users are using the newest spikeinterface.

**Comment by magland** [comment]
Thanks for the report @jbmelander 

Not sure if this is it, but one thing that changed was how the binary (cached) recording is created. See this commit from 6 months ago

https://github.com/SpikeInterface/spikeinterface/commit/51b8dcedd2b9d9de404ea212250a265e4ad130be

Are you doing the filtering and whitening before passing into spikeinterface? What type of recording extractor are you passing in?

Tagging @alejoe91

**Comment by jbmelander** [comment]
So, I do the filtering and whitening as a preprocessing step before caching it. Then I load the cached recording and use that for ms5. I found this to be much faster than just using the lazy preprocessing. Could this have something to do with it?

For reference, here is the [ms5 quickstart guide](https://github.com/flatironinstitute/mountainsort5/blob/main/examples/neuropixel_quickstart/spikeglx.py) (which I will update today, if we can identify the bug)

**Comment by magland** [comment]
Could you please try using create_cached_recording instead of save_binary_recording (it's also in mountainsort5.util)

**Comment by alejoe91** [comment]
@jbmelander are you setting `n_jobs` to -1 either globally or when running MS5?

**Comment by alejoe91** [comment]
Actually, you should use this parameter for MS5: `n_jobs_for_preprocessing=-1`

We should add that to the tutorial! Can you open an issue on SI?

**Comment by magland** [comment]
@alejoe91 I'm thinking that perhaps @jbmelander is not calling ms5 via SI... this is based on looking at [his script](https://github.com/flatironinstitute/mountainsort5/blob/main/examples/neuropixel_quickstart/spikeglx.py). But somehow the newer SI is causing problems elsewhere....? ... unless I am misunderstanding what's going on

**Comment by jbmelander** [comment]
I'm looking into it still. I will post back here with a report on what I've found later today.

But yes, @alejoe91 @magland is correct that I'm not calling ms5 sorting itself from SI.
Thanks everyone for the quick replies.

**Comment by alejoe91** [comment]
Nothing comes to mind. @jbmelander have you identified specific steps that became slower?

With the latest spikeinterface version, the `extract_waveforms` is deprecated in favor of the `SortingAnalyzer` approach. See [this guide](https://spikeinterface.readthedocs.io/en/latest/tutorials/waveform_extractor_to_sorting_analyzer.html) for how to update the script

**Comment by jbmelander** [comment]
The problem is occurring way before the extract_waveforms step. Yes, there are several steps that are either slower or slower AND broken. I have a few hypotheses about what's going on but will wait to share until I've finished running a few tests. Thanks!

**Comment by jbmelander** [comment]
I am pretty baffled. 
If i use preprocess my recording with whiten and bandpass, then pass that into ms5.sorting_scheme_2(), there are dramatic changes in speed and performance. However, if I run it via si.run_sorter(...) performance is great. 
I will keep you updated, hoping to resolve this today.

**Comment by jbmelander** [comment]
Alright, solved. 
For both my binary recording load/save util and for the ms5.utils.create_cached_recording, if I whiten and filter before caching, the newer SI version returns the unwhitened and unfiltered data when the recording is loaded. Does this sound like it might make sense to you @alejoe91 @magland?

**Comment by jbmelander** [comment]
I should add - the reason I was doing the whitening and filtering prior to caching, is that I had found that ms5 ran faster on the cached pre-processed data than it did using the "lazy" scheme. The slowdown and performance drop I noticed with newer SI versions are equivalent to running ms5 without any filtering using the older SI version (0.99.1) that we have been using for a while. I also notice that it seems like there is some sort of caching occurring in the sorters/external/mountainsort5 implementation, so I'm curious:

1. Why that caching is different than the implementations in ms5/utils
2. If you could explain what the breaking change was between 0.99.1 and the latest SI version

Thanks! Either way, its working again which is the important thing, and I will update the spikeglx quickstart in ms5 this weekend.

**Comment by jbmelander** [comment]
Does this make sense to you, @alejoe91 ?

**Comment by alejoe91** [comment]
Yes. The cached data is whitened and filtered, but it doesn't know! So in that case you have to manually set filter and whiten to False. Another option when running through spikeinterface is to not filter and whiten before, since the wrapper will automatically do these ops and cache to binary!
